jOOL-java-8/src/main/java/org/jooq/lambda/Agg.java

```diff
@@ -906,11 +906,27 @@ else if (l1[0] == null)
 
     /**
      * Get a {@link Collector} that calculates the derived <code>PERCENTILE_DISC(percentile)</code> function given a specific ordering.
+     *
+     * @param function map the items in the streams into values
+     * @param comparator comparator used for sorting the items
+     * @return a collector that calculates the derived <code>PERCENTILE_DISC(percentile)</code> function
      */
     public static <T, U> Collector<T, ?, Optional<T>> percentileBy(double percentile, Function<? super T, ? extends U> function, Comparator<? super U> comparator) {
         if (percentile < 0.0 || percentile > 1.0)
             throw new IllegalArgumentException("Percentile must be between 0.0 and 1.0");
 
+        // CS304 Issue link: https://github.com/jOOQ/jOOL/issues/376
+        if (percentile == 0.0)
+            // If percentile is 0, this is the same as taking the item with the minimum value.
+            return minBy(function, comparator);
+        else if (percentile == 1.0)
+            // If percentile is 1, this is the same as taking the item with the maximum value,
+            // If there are multiple maxima, take the last one.
+            return maxBy(function, (o1, o2) -> {
+                int compareResult = comparator.compare(o1, o2);
+                return compareResult == 0 ? -1 : compareResult;
```

Comment:
This hack violates the `Comparator` contract. We can't implement it like this.

Reply:
This seems to be the correct way to implement this:

```java
collectingAndThen(maxAllBy(function, comparator), s -> s.findLast())
```

Reply:
Emmm, I think my solution is not incorrect, because the function `maxBy` will not sort the elements but compare them one by one, and the input comparator will not be modified. So even though this implementation violates the design philosophy of `Comparator`, it works and there seems no potential problem with it.

It's perfectly fine to implement this with `maxAllBy`, but in the worst case the space complexity will be O(n). That's my concern.

Reply:
It's incorrect. If `comparator.compare(o1, o2) == 0`, then `comparator.compare(o2, o1) == 0`, yet you return `-1` in both cases. I don't want to spend the time now to find an edge case where this breaks sorting algorithms, but it should be easy to get an intuition about how this hack feels very wrong

> because the function `maxBy` will not sort the elements but compare them one by one

You shouldn't rely on such an implementation detail.

> So even though this implementation violates the design philosophy of `Comparator`, it works and there seems no potential problem with it.

Famous last words :)

> It's perfectly fine to implement this with `maxAllBy`, but in the worst case the space complexity will be O(n). That's my concern.

I'm open to other suggestions, but correctness always beats performance.

Reply:
Yes, you are right. Thanks for your reply!